aspnet/6-pubsub/lib/BookDetailLookup.cs

```diff
@@ -223,15 +225,61 @@ public static void ProcessBook(IBookStore bookStore, long bookId)
             var response = WebRequest.Create(query).GetResponse();
             var reader = new StreamReader(response.GetResponseStream());
             var json = reader.ReadToEnd();
-            dynamic results = JObject.Parse(json);
-            dynamic info = results.items[0].volumeInfo;
-            book.Title = info.title;
-            book.PublishedDate = info.publishedDate;
-            book.Author = string.Join(", ", info.authors);
-            book.Description = info.description;
+            ParseBook(json, ref book);
             // TODO: update image.
             bookStore.Update(book);
         }
         // [END processbook]
+
+        /// <summary>
+        /// Parse a date time.  Return null if it can't be parsed.  A single number will be 
+        /// interpreted as a year, and the date returned will be YEAR-01-01.
+        /// </summary>
+        /// <param name="dateString">A string representation of the date.</param>
+        /// <returns>A datetime or null.</returns>
+        public static DateTime? ParseDate(string dateString)
+        {
+            DateTime result;
+            if (DateTime.TryParse(dateString, out result))
+                return result;
+            int year;
+            if (int.TryParse(dateString, out year))
+                return new DateTime(year, 1, 1);
+            return null;
+        }
+
+        /// <summary>
+        /// Updates book with information parsed from a json response from Google's Books API.
+        /// </summary>
+        /// <param name="json">A response from Google's Books API.</param>
+        /// <param name="book">Fields in book will be overwritten.</param>
+        public static void ParseBook(string json, ref Book book)
+        {
+            // There are many volumeInfos, and many are incomplete.  So, to find a field like
+            // "title", we use LINQ to scan multiple volumeInfos.
+            JObject results = JObject.Parse(json);
+            var infos = results["items"].Select(token => (JObject)token)
+                .Where(obj => obj.Property("volumeInfo") != null)
+                .Select(obj => obj["volumeInfo"]);
+            Func<string, IEnumerable<JToken>> GetInfo = (propertyName) =>

```

Comment:
> If the final volumeInfo does not contain a title or description, then book.Title or book.Description will be set to null.

Is the overarching goal of this code to find more *compete* book results that contain a minimum set of data?  For example: for the purposes of this application, a result should ideally contain a least a title, author, and book image.

Can we make it more clear in the code that this is the goal?  For example, this code demonstrates parsing a JSON response from the Books API and, from the response, finding book volumes that contain at least a title, author, and book image:

```cs
void UpdateBookDetails(string json, ref Book book)
{
    var bookApiResults = JObject.Parse(json);
    
    var bookResult = bookApiResults["items"].
      Select(item => item["volumeInfo"]).
      Where(v => v["title"] != null && v["authors"] != null && v["image_links"]?["thumbnail"] != null).
      FirstOrDefault();

    if (bookResult != null)
    {
        book.Title       = bookResult["title"].ToString();
        book.Description = bookResult["description"]?.ToString();
        book.Author      = string.Join(", ", bookResult["authors"]);

        if (bookResult["publishedDate"] != null)
        {
            var publishedDate = bookResult["publishedDate"].ToString();
            
            // Published date can be in the format 2010 or 2010-01-01
            if (publishedDate.Length == 4)
                publishedDate = $"{publishedDate}-01-01"
            
            book.PublishedDate = DateTime.Parse(info["publishedDate"]);
        }
    }
}
```